### PR TITLE
Atrous (Dilated) Convolution Implementation

### PR DESCRIPTION
**Pull request by ShikharJ**

### CODE REVIEW

File: src/mlpack/methods/ann/layer/atrous_convolution.hpp [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+#include <mlpack/methods/ann/convolution_rules/svd_convolution.hpp>
+
+#include "layer_types.hpp"
+
+namespace mlpack{
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Implementation of the Atrous Convolution class. The Atrous Convolution
+ * class represents a single layer of a neural network.
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/atrous_convolution.hpp]: Would it be possible to add a little bit about what atrous convolution is and why someone would want it here?

File: src/mlpack/methods/ann/layer/atrous_convolution_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+
+    outputTemp.slice(outMap) += bias(outMap);
+  }
+
+  outputWidth = outputTemp.n_rows;
+  outputHeight = outputTemp.n_cols;
+
+  outputTemp.reshape(outputTemp.n_elem, 1, 1);
+  output = std::move(outputTemp.slice(0));
+  std::cout << output << std::endl;
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/atrous_convolution_impl.hpp]: Looks like some extra debugging output here that we should probably remove.

File: src/mlpack/methods/ann/layer/atrous_convolution_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+    }
+
+    outputTemp.slice(outMap) += bias(outMap);
+  }
+
+  outputWidth = outputTemp.n_rows;
+  outputHeight = outputTemp.n_cols;
+
+  outputTemp.reshape(outputTemp.n_elem, 1, 1);
+  output = std::move(outputTemp.slice(0));
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/atrous_convolution_impl.hpp]: If the idea of setting `outputTemp` to point to the same memory as `output` isn't feasible, I think this is a fine way to do it.

**Comment by ShikharJ** [comment]
This is ready for a review @zoq @rcurtin.

**Comment by ShikharJ** [comment]
@rcurtin @zoq Sorry for being impatient, but the changes here are required for debugging the `GAN` PR. Could we get this merged `ASAP`?

File: src/mlpack/methods/ann/convolution_rules/naive_convolution.hpp [hunk 4]
```diff
@@ -91,20 +99,39 @@ class NaiveConvolution
...
+              const size_t dilationH = 1)
   {
-    const size_t outputRows = (input.n_rows + 2 * (filter.n_rows - 1)) * dW;
-    const size_t outputCols = (input.n_cols + 2 * (filter.n_cols - 1)) * dH;
+    size_t outputRows = (input.n_rows - 1) * dW + 2 * (filter.n_rows - 1)
+        * dilationW + 1;
+    size_t outputCols = (input.n_cols - 1) * dH + 2 * (filter.n_cols - 1)
+        * dilationH + 1;
+
+    for (size_t i = 0; i < dW; i++){
```
Comment (zoq) [inline src/mlpack/methods/ann/convolution_rules/naive_convolution.hpp]: Pedantic style, can you put the opening ``{`` on a new line.

File: src/mlpack/methods/ann/layer/atrous_convolution_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+      arma::Cube<eT> output, reducedOutput;
+      GradientConvolutionRule::Convolution(inputSlices, deltaSlices,
+          output, dW, dH, 1, 1);
+      arma::Mat<eT> reducedMat;
+      reducedOutput = arma::zeros<arma::Cube<eT> >(kW, kH, output.n_slices);
+
+      for (size_t j = 0; j < output.n_slices; j++)
+      {
+        reducedMat = output.slice(j);
+        if (dilationH != 1){
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/atrous_convolution_impl.hpp]: See comment above, also maybe ``> 1`` is more descriptive here.

**Comment by zoq** [comment]
I guess, an easy option would be to copy the files in your local repo and don't commit the files, once this is merged you could rebase the branch against the master branch and use that.

**Comment by zoq** [comment]
Thanks for another really great contribution.

**Pull request merged**